File: Sources/Golos/HotkeyOption.swift
import Foundation

/// Клавиша, по которой идёт запись.
///
/// Список закрытый намеренно. Свободный захват «нажми любую клавишу» позволяет
/// назначить букву — и человек остаётся без клавиатуры, потому что каждое её
/// нажатие уходит в диктовку. Здесь только те клавиши, что в обычном наборе
/// не участвуют.
struct HotkeyOption: Identifiable, Hashable {
    let id: String
    let title: String
    /// Код клавиши в событиях CGEvent.
    let keyCode: Int64
    /// Бит в маске флагов — только у модификаторов. У обычных клавиш nil:
    /// они приходят не через flagsChanged, а через keyDown и keyUp.
    let flagMask: UInt64?
    let note: String

    var isModifier: Bool { flagMask != nil }

    init(id: String, title: String, keyCode: Int64, flagMask: UInt64?, note: String) {
        self.id = id
        self.title = title
        self.keyCode = keyCode
        self.flagMask = flagMask
        self.note = note
    }

    static let all: [HotkeyOption] = [
        HotkeyOption(id: "rightOption", title: "Правый ⌥", keyCode: 61, flagMask: 0x0000_0040,
                     note: "По умолчанию. В наборе не участвует"),
        HotkeyOption(id: "rightCommand", title: "Правый ⌘", keyCode: 54, flagMask: 0x0000_0010,
                     note: "Свободен, если не пользуешься сочетаниями правой рукой"),
        HotkeyOption(id: "rightControl", title: "Правый ⌃", keyCode: 62, flagMask: 0x0000_2000,
                     note: "Свободен на большинстве клавиатур"),
        HotkeyOption(id: "rightShift", title: "Правый ⇧", keyCode: 60, flagMask: 0x0000_0004,
                     note: "Осторожно: им набирают заглавные"),
        HotkeyOption(id: "f13", title: "F13", keyCode: 105, flagMask: nil,
                     note: "Есть на полноразмерных клавиатурах"),
        HotkeyOption(id: "f14", title: "F14", keyCode: 107, flagMask: nil,
                     note: "Есть на полноразмерных клавиатурах"),
        HotkeyOption(id: "f15", title: "F15", keyCode: 113, flagMask: nil,
                     note: "Есть на полноразмерных клавиатурах"),
        HotkeyOption(id: "f16", title: "F16", keyCode: 106, flagMask: nil,
                     note: "Есть на полноразмерных клавиатурах"),
        // нажатие и отпускание, удержание в три секунды доходит целиком.
    ]

    static let fallback = all[0]

    static func named(_ id: String) -> HotkeyOption {
        all.first { $0.id == id } ?? fallback
    }
}

File: windows/src/HotkeyOption.cs
namespace Golos;

/// <summary>
/// Клавиша, по которой идёт запись.
///
/// Список закрытый намеренно. Свободный захват «нажми любую клавишу» позволяет
/// назначить букву — и человек остаётся без клавиатуры, потому что каждое её
/// нажатие уходит в диктовку.
///
/// Идентификаторы общие с версией для macOS: там правый Alt называется
/// правым Option, но это одна и та же физическая клавиша, и конфиг должен
/// переноситься между машинами без правок.
/// </summary>
public sealed record HotkeyOption(string Id, string Title, int VirtualKey, string Note)
{
    public static readonly HotkeyOption[] All =
    {
        new("rightOption",  "Правый Alt",     0xA5, "По умолчанию. В наборе не участвует"),
        new("rightCommand", "Правый Windows", 0x5C, "Свободен, если не пользуешься меню «Пуск» с клавиатуры"),
        new("rightControl", "Правый Ctrl",    0xA3, "Свободен на большинстве клавиатур"),
        new("rightShift",   "Правый Shift",   0xA1, "Осторожно: им набирают заглавные"),
        new("f13",          "F13",            0x7C, "Есть на полноразмерных клавиатурах"),
        new("f14",          "F14",            0x7D, "Есть на полноразмерных клавиатурах"),
        new("f15",          "F15",            0x7E, "Есть на полноразмерных клавиатурах"),
        new("f16",          "F16",            0x7F, "Есть на полноразмерных клавиатурах")
    };

    public static HotkeyOption Fallback => All[0];

    public static HotkeyOption Named(string id) =>
        All.FirstOrDefault(o => o.Id == id) ?? Fallback;

    /// <summary>
    /// Модификаторы проглатываем: одиночное нажатие Alt или Windows отдаёт
    /// фокус меню, а Shift и Ctrl сами по себе ничего не делают, но их
    /// сочетания ломать нельзя — поэтому глотаем только когда клавиша
    /// нажата одна.
    /// </summary>
    public bool IsModifier => VirtualKey is 0xA5 or 0x5C or 0xA3 or 0xA1;
}
